apps/balancer-tools/src/app/apr/(utils)/tokenApr.ts

```diff
@@ -285,3 +321,159 @@ const getImplementationAddress = async (
+  timeEnd: number;
+}) {
+  const rangeData = await poolsWithCache.gql(chain).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(timeStart, timeEnd),
+  });
+
+  let pool: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    pool = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(chain).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    pool = currentPoolData.pool;
+  }
+
+  if (!pool || !pool.tokens) {
+    return 0;
+  }
+
+  let apr;
+
+  if (
+    pool.poolType === PoolTypeEnum.PHANTOM_STABLE ||
+    (pool.poolType === PoolTypeEnum.WEIGHTED &&
+      (pool.poolTypeVersion ?? 0) >= 2)
+  ) {
+    const isExemptFromYieldProtocolFee = pool?.tokens.find(
+      ({ address }) => address === tokenAddress,
+    )?.isExemptFromYieldProtocolFee;
+
+    if (isExemptFromYieldProtocolFee) {
+      apr = tokenYield;
+    } else {
+      //source: https://github.com/balancer/balancer-sdk/blob/f4879f06289c6f5f9766ead1835f4f4b096ed7dd/balancer-js/src/modules/pools/apr/apr.ts#L134
+      apr = tokenYield * (1 - pool.protocolYieldFeeCache || 0.5);
+    }
+  } else if (
+    pool.poolType === PoolTypeEnum.META_STABLE ||
+    pool.poolType?.includes("Gyro")
+  ) {
+    apr = tokenYield * (pool.protocolYieldFeeCache || 0.5);
+  }
+
+  apr = (apr || 0) * tokenWeight;
+
+  return apr;
+}
+
+async function getTokenWeight(
+  tokenAddress: string,
+  startAtTimestamp: number,
+  endAtTimestamp: number,
+  poolNetwork: string,
+  poolId: string,
+): Promise<number> {
+  const rangeData = await poolsWithCache.gql(poolNetwork).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(startAtTimestamp, endAtTimestamp),
+  });
+
+  let poolData: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    poolData = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(poolNetwork).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    poolData = currentPoolData.pool;
+  }
+
+  const poolTokenData = poolData.tokens?.filter((token) => {
+    return token.address !== poolData.address;
+  });
+
+  if (!Array.isArray(poolTokenData)) {
+    return 0;
+  }
+
+  const token = poolTokenData.find((t) => t.address === tokenAddress);
+  if (!token) {
+    // eslint-disable-next-line no-console
+    console.warn(`Token not found in pool for address: ${tokenAddress}`);
+    return 0;
+  }
+
+  if (poolData.poolType === PoolTypeEnum.WEIGHTED && token.weight) {
+    return parseFloat(token.weight);
+  }
+
+  const tokensPrices = await Promise.all(
+    poolTokenData.map(async (token) => {
+      if (
+        vunerabilityAffecteRateProviders.some(
+          ({ address }) =>
+            address.toLowerCase() === token.address.toLowerCase(),
+        )
+      ) {
+        return 1;
```

Comment:
why is 1 the default?

Reply:
actually, I was thinking about this, how not to blow the token weight when one of the tokens doesn't have their price on coinllama